packages/nextjs-kit/src/components/paginator.tsx

```diff
@@ -0,0 +1,257 @@
+  breakpoints,
+  itemsPerPage,
+  routing,
+  Component,
+}: PaginationProps) => {
+  // configurable breakpoints
+  // This value will be the start of the separator.
+  const [breakStart, setBreakStart] = useState<number | null>(
+    breakpoints?.start || null
+  );
+
+  // This value will be the button to start with after separator
+  const breakEnd = breakpoints?.end || null;
+  // how many buttons to add when the separator is clicked
+  const breakAdd = breakpoints?.add || null;
+
+  const router = useRouter();
+  // get current path from router.pathname
+  // and trim off catchalls
+  const currentRoute = router.pathname.replace(/\/{1}\[{1,2}.*\]{1,2}$/, '');
+
+  const [currentPageQuery, setCurrentPageQuery] = useState<number>(
+    Number(router.query.page) || 1
+  );
+
+  const [offset, setOffset] = useState<number>(
+    (currentPageQuery - 1) * itemsPerPage
+  );
+  const [currentItems, setCurrentItems] = useState<Array<PaginatorObj>>([]);
+  const [totalItems] = useState<number>(data.length);
+
+  const [totalPages] = useState<number>(Math.ceil(data.length / itemsPerPage));
+
+  // since we fetch ALL items, filtering the current items client side
+  // further reduces API calls to the server
+  useEffect(() => {
+    setCurrentItems(
+      data.slice(
+        (currentPageQuery - 1) * itemsPerPage,
+        itemsPerPage * currentPageQuery
+      )
+    );
+    // set the new offset
+    const newOffset = ((currentPageQuery - 1) * itemsPerPage) % totalItems;
+    if (currentPageQuery > 0) {
+      setOffset(Number(newOffset));
+    }
+
+    routing &&
+      router.push(
+        `${currentRoute}/${currentPageQuery}`,
+        `${currentRoute}/${currentPageQuery}`,
+        {
+          shallow: true,
+        }
+      );
+  }, [data, offset, itemsPerPage, breakStart, currentPageQuery, totalItems]);
+
+  // track window width to appropriately hide and show buttons on small viewports
+  const [windowWidth, setWindowWidth] = useState<number>();
+  useEffect(() => {
+    setWindowWidth(window.innerWidth);
+    window.addEventListener('resize', () => {
+      setWindowWidth(window.innerWidth);
+    });
+  }, []);
+  // Render the currentItems to the page
+  const RenderData = () => {
+    return (
+      <React.Fragment>
+        {currentItems.map(data => {
+          return (
+            <article
+              key={data.id}
+              className="flex flex-col p-3 w-fit mx-auto mb-10"
+            >
+              {data?.title && (
+                <h2 className="justify-start my-auto text-2xl mb-2">
+                  {data.title}
+                </h2>
+              )}
+              {data?.body?.value && (
+                <p className="max-w-prose my-2">
+                  {data?.body.value.substring(0, 150)}...
+                </p>
+              )}
+            </article>
+          );
+        })}
+      </React.Fragment>
+    );
+  };
+
+  const handlePageClick: React.MouseEventHandler<HTMLButtonElement> = event => {
```

Comment:
here is another potential good case for a type predicate to check if the event.target is an instance of an Element, instead of relying on `as`, which can be troublesome.

Reply:
I used `as` here not to check if event.target is an instance of element but rather because the EventTarget type that we pass in does not automatically inherit from the Element type. [Here](https://www.designcise.com/web/tutorial/how-to-fix-property-does-not-exist-on-type-eventtarget-typescript-error) is a good blog post I found explaining the issue but I think it may be fine to use `as` in this case? Let me know your thoughts.

Reply:
True, but we could use a type predicate to check if something is an HTML element and avoid the type coercin, since we will know forsure what type we have.
example:

```typescript
// type predicate that determines an element is an HTMLElement:
  const isHTMLElement = (element: unknown): element is HTMLElement => {
    return element instanceof HTMLElement;
  };
```

Then in the `handlePageClick`:

```typescript
// return early if we don't have the right element type
if (!isHTMLElement(event.target)) return;
  // this should not throw typescript errors because it knows it is an HTMLElement at this point.
    if (event.target.id === 'next-btn') {
...
```